src/client/Microsoft.Identity.Client/Utils/OptionalSemaphoreSlim.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Utils
+{
+    /// <summary>
+    /// An object that either wraps a SemaphoreSlim for syncronization or ignores syncronization completely and just keeps track of Wait / Release operations.
+    /// </summary>
+    internal class OptionalSemaphoreSlim
+    {
+        private readonly bool _useRealSemaphore;
+        private int _noLockCurrentCount;
+        private SemaphoreSlim _semaphoreSlim;
+
+        public int CurrentCount
+        {
+            get
+            {
+                return _useRealSemaphore ? _semaphoreSlim.CurrentCount : _noLockCurrentCount;
+            }
+        }
+
+        public OptionalSemaphoreSlim(bool useRealSemaphore)
+        {
+            _useRealSemaphore = useRealSemaphore;
+            if (_useRealSemaphore)
+            {
+                _semaphoreSlim = new SemaphoreSlim(1, 1);
+            }
+            _noLockCurrentCount = 1;
+        }
+
+        public void Release()
+        {
+            if (_useRealSemaphore)
+            {
+                _semaphoreSlim.Release();
+            }
+            else
+            {
+                Interlocked.Increment(ref _noLockCurrentCount);
```

Comment:
What is the benefit of using Interlocked to increment/decrement the int?

Reply:
This is because an increment is actually 2 operations

```csharp
i++; // this is quick way of writing...

j = i + 1; // increment
i = j;       // assign
```

In a multi-threaded environment, a race condition can occur where 1 thread executes the increment, but not the assign, and thread 2 executes both. 

Interlocked is just a faster way of doing

```csharp
lock (obj)
{
    i++;
}

```


This blog describes it in more detail: https://tim.silvrback.com/dot-net-interlocked-atomic-operations